Source/Core/VideoCommon/VertexManagerBase.cpp

```diff
@@ -235,9 +263,69 @@ void VertexManager::Flush()
 		ERROR_LOG(VIDEO, "xf.numtexgens (%d) does not match bp.numtexgens (%d). Error in command stream.", xfmem.numTexGen.numTexGens, bpmem.genMode.numtexgens.Value());
 
 	IsFlushed = true;
+	CullAll = false;
 }
 
 void VertexManager::DoState(PointerWrap& p)
 {
+	p.Do(ZSlope);
 	g_vertex_manager->vDoState(p);
 }
+
+void VertexManager::CalculateZSlope(NativeVertexFormat *format)
+{
+	float vtx[9];
+	float out[12];
+	float viewOffset[2] = { xfmem.viewport.xOrig - bpmem.scissorOffset.x * 2,
+	                        xfmem.viewport.yOrig - bpmem.scissorOffset.y * 2};
+
+	// Global matrix ID.
+	u32 mtxIdx = g_main_cp_state.matrix_index_a.PosNormalMtxIdx;
+	const PortableVertexDeclaration vert_decl = format->GetVertexDeclaration();
+	size_t posOff = vert_decl.position.offset;
+	size_t mtxOff = vert_decl.posmtx.offset;
+
+	// Lookup vertices of the last rendered triangle and software-transform them
+	// This allows us to determine the depth slope, which will be used if z--freeze
+	// is enabled in the following flush.
+	for (unsigned int i = 0; i < 3; ++i)
+	{
+		u8* vtx_ptr = s_pCurBufferPointer - vert_decl.stride * (3 - i);
```

Comment:
There is a potential for a buffer underrun here if the buffer contains less than 3 vertices (say a single line or less than 3 points.)